// oip-react/lib/components/ModuleWallet/ActionNavBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactJss = _interopRequireDefault(require("react-jss"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = theme => ({
  root: {
    display: 'flex',
    flexDirection: 'row',
    flex: '1 1 auto',
    justifyContent: 'flex-start',
    alignItems: 'center'
  },
  navItemContainer: {
    display: 'flex',
    flexDirection: 'row',
    flex: '1',
    padding: [5, 10],
    cursor: 'pointer',
    justifyContent: 'center'
  }
});

const ActionNavBar = ({
  classes,
  navItems = [],
  activeNavLink,
  onNavLinkClick,
  theme
}) => {
  function activeNavLinkStyle(item) {
    if (item === activeNavLink) {
      return {
        borderBottom: `2px solid ${theme.palette.primary.main}`
      };
    }
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: classes.root
  }, navItems.map((item, i) => {
    return /*#__PURE__*/_react.default.createElement("div", {
      key: i,
      className: classes.navItemContainer,
      onClick: () => onNavLinkClick(item),
      style: activeNavLinkStyle(item)
    }, /*#__PURE__*/_react.default.createElement("span", null, item));
  }));
};

var _default = (0, _reactJss.default)(styles, {
  injectTheme: true
})(ActionNavBar);

exports.default = _default;
