// oip-react/lib/components/ModuleWallet/Addresses.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactJss = _interopRequireDefault(require("react-jss"));

var _clipboard = _interopRequireDefault(require("clipboard"));

var _classnames = _interopRequireDefault(require("classnames"));

var _icons = require("@material-ui/icons");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = theme => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    flex: '10',
    overflowY: 'auto'
  },
  addressContainer: {
    display: 'flex',
    flexDirection: 'row',
    flex: '0 0 30px',
    fontFamily: 'monospace',
    alignItems: 'center',
    marginRight: [5, 5],
    padding: [0, 7],
    '& span': {
      justifyContent: 'center'
    }
  },
  addressIndex: {
    display: 'flex',
    flex: '1'
  },
  pubAddress: {
    display: 'flex',
    flex: '9',
    cursor: 'copy',
    '&:hover': {
      color: 'blue'
    }
  },
  explorerLink: {
    display: 'flex',
    flex: '1',
    cursor: 'pointer',
    '&:hover': {
      color: 'red'
    }
  },
  copyWIF: {
    display: 'flex',
    flex: '1',
    cursor: 'copy',
    '&:hover': {
      color: 'orange'
    }
  },
  addAddress: {
    border: '1px solid grey',
    padding: '5px 7px',
    borderRadius: '3px',
    fontSize: '12px',
    marginLeft: '15px',
    cursor: 'pointer'
  },
  addRow: {
    margin: '15px 0px'
  }
});

const Addresses = ({
  classes,
  addAddress,
  addresses,
  explorerUrl
}) => {
  new _clipboard.default('.copy-p2pkh');
  explorerUrl = explorerUrl.split('api')[0];
  return /*#__PURE__*/_react.default.createElement("div", {
    className: classes.root
  }, addresses.map(addr => {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: classes.addressContainer,
      key: addr.address.index
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: classes.addressIndex
    }, "/", addr.address.index), /*#__PURE__*/_react.default.createElement("span", {
      "data-clipboard-target": `#addr-${addr.address.index}`,
      id: `addr-${addr.address.index}`,
      className: (0, _classnames.default)('copy-p2pkh', classes.pubAddress),
      onClick: () => {
        const sel = window.getSelection();
        sel.empty();
        window.alert('copied address to clipboard');
      }
    }, addr.getPublicAddress()), /*#__PURE__*/_react.default.createElement("span", {
      className: classes.explorerLink,
      onClick: () => window.open(`${explorerUrl}/address/${addr.getPublicAddress()}`, '_blank')
    }, /*#__PURE__*/_react.default.createElement(_icons.Link, null)), /*#__PURE__*/_react.default.createElement("span", {
      className: classes.copyWIF,
      onClick: () => {
        navigator.clipboard.writeText(addr.getPrivateAddress());
        window.alert('Copied Private Key in WIF format to clipboard');
      }
    }, /*#__PURE__*/_react.default.createElement(_icons.Fingerprint, null)));
  }), addresses.length > 0 ? /*#__PURE__*/_react.default.createElement("div", {
    className: classes.addRow
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: classes.addAddress,
    onClick: () => addAddress()
  }, "+ Show next address")) : null);
};

var _default = (0, _reactJss.default)(styles)(Addresses);

exports.default = _default;
