// oip-react/lib/components/ModuleWallet/ModuleWallet.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactJss = _interopRequireDefault(require("react-jss"));

var _hdmw = require("@oipwg/hdmw");

var _styledJss = _interopRequireDefault(require("styled-jss"));

var _index = require("./index");

var _WalletHeader = _interopRequireDefault(require("./WalletHeader"));

var _WalletStateContainer = _interopRequireDefault(require("./WalletStateContainer"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const styles = theme => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    flex: '1 1 600px',
    borderRadius: 'inherit',
    backgroundColor: 'white'
  }
});

const Container = (0, _styledJss.default)('div')({
  width: props => props.width ? props.width : '779px',
  height: props => props.height ? props.height : '339px',
  borderRadius: props => props.borderRadius ? props.borderRadius : '10px',
  display: 'flex',
  flexDirection: 'row',
  flexGrow: props => props.flexGrow,
  flexShrink: props => props.flexShrink || '0',
  flex: props => props.flex,
  alignSelf: props => props.alignSelf,
  justifySelf: props => props.justifySelf,
  border: props => props.border,
  boxShadow: props => props.boxShadow
});

const ModuleWallet = ({
  classes,
  coins,
  ...rest
}) => {
  const walletRef = (0, _react.useRef)(null);
  const [lock, setLock] = (0, _react.useState)(true);

  function onMnemonicSubmit(mnemonic) {
    walletRef.current = new _hdmw.Wallet(mnemonic, {
      discover: false,
      supportedCoins: coins
    });
    setLock(false);
  }

  const resetWallet = () => {
    walletRef.current = null;
    setLock(true);
  };

  return /*#__PURE__*/_react.default.createElement(Container, rest, /*#__PURE__*/_react.default.createElement("div", {
    className: classes.root
  }, /*#__PURE__*/_react.default.createElement(_WalletHeader.default, {
    resetWallet: resetWallet
  }), lock ? /*#__PURE__*/_react.default.createElement(_index.LoadWallet, {
    onMnemonicSubmit: onMnemonicSubmit
  }) : /*#__PURE__*/_react.default.createElement(_WalletStateContainer.default, {
    wallet: walletRef.current,
    coins: coins
  })));
};

var _default = (0, _reactJss.default)(styles)(ModuleWallet);

exports.default = _default;
