// oip-react/lib/components/ModuleWallet/Settings.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactJss = _interopRequireDefault(require("react-jss"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = theme => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    flex: '1',
    overflowY: 'auto'
  },
  settingsContainer: {
    margin: [10, 20]
  },
  settingsTitle: {
    marginTop: 0,
    marginBottom: 5,
    color: 'grey',
    fontFamily: 'arial'
  }
});

const Settings = ({
  classes,
  coins,
  wallet
}) => {
  return /*#__PURE__*/_react.default.createElement("div", {
    className: classes.root
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: classes.settingsContainer
  }, /*#__PURE__*/_react.default.createElement("h4", {
    className: classes.settingsTitle
  })));
};

var _default = (0, _reactJss.default)(styles)(Settings);

exports.default = _default;
