// oip-react/lib/components/ModuleWallet/WalletPages.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactJss = _interopRequireDefault(require("react-jss"));

var _Addresses = _interopRequireDefault(require("./Addresses"));

var _Transactions = _interopRequireDefault(require("./Transactions"));

var _Send = _interopRequireDefault(require("./Send"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = theme => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    flex: '9'
  }
});

const ADDRESSES = 'Addresses';
const TRANSACTIONS = 'Transactions';
const SEND = 'Send';

const WalletPages = ({
  classes,
  wallet,
  activeCoin,
  activeNavLink,
  state,
  handleAddAddress
}) => {
  const addresses = activeNavLink === ADDRESSES;
  const transactions = activeNavLink === TRANSACTIONS;
  const send = activeNavLink === SEND;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: classes.root
  }, addresses && /*#__PURE__*/_react.default.createElement(_Addresses.default, {
    addresses: state[activeCoin].addresses,
    addAddress: handleAddAddress,
    explorerUrl: wallet.getExplorerUrls()[activeCoin]
  }), transactions && /*#__PURE__*/_react.default.createElement(_Transactions.default, {
    transactions: state[activeCoin].transactions,
    explorerUrl: wallet.getExplorerUrls()[activeCoin],
    addresses: state[activeCoin].addresses
  }), send && /*#__PURE__*/_react.default.createElement(_Send.default, {
    activeCoin: activeCoin,
    wallet: wallet
  }));
};

var _default = (0, _reactJss.default)(styles)(WalletPages);

exports.default = _default;
