// oip-react/lib/components/ModuleWallet/WalletBody.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactJss = _interopRequireDefault(require("react-jss"));

var _ActionNavBar = _interopRequireDefault(require("./ActionNavBar"));

var _WalletPages = _interopRequireDefault(require("./WalletPages"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = theme => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    flex: '9',
    overflow: 'auto'
  }
});

const WalletBody = ({
  classes,
  wallet,
  onNavLinkClick,
  navItems,
  activeNavLink,
  activeCoin,
  coins,
  state,
  handleAddAddress
}) => {
  return /*#__PURE__*/_react.default.createElement("div", {
    className: classes.root
  }, /*#__PURE__*/_react.default.createElement(_ActionNavBar.default, {
    navItems: navItems,
    onNavLinkClick: onNavLinkClick,
    activeNavLink: activeNavLink
  }), /*#__PURE__*/_react.default.createElement(_WalletPages.default, {
    activeCoin: activeCoin,
    wallet: wallet,
    coins: coins,
    activeNavLink: activeNavLink,
    state: state,
    handleAddAddress: handleAddAddress
  }));
};

var _default = (0, _reactJss.default)(styles)(WalletBody);

exports.default = _default;
