// oip-react/lib/components/ModuleWallet/WalletHeader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactJss = _interopRequireDefault(require("react-jss"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = theme => ({
  root: {
    display: 'flex',
    flexDirection: 'row',
    flex: '1 1 30px',
    backgroundColor: `${theme.palette.primary.main}`,
    borderTopRightRadius: 'inherit',
    borderTopLeftRadius: 'inherit',
    overflow: 'hidden',
    justifyContent: 'space-between',
    padding: [0, 10],
    alignItems: 'center',
    color: 'white'
  }
});

const WalletHeader = ({
  classes,
  resetWallet
}) => {
  return /*#__PURE__*/_react.default.createElement("div", {
    className: classes.root
  }, /*#__PURE__*/_react.default.createElement("h5", null, "hdmw-module"), /*#__PURE__*/_react.default.createElement("span", {
    onClick: resetWallet,
    style: {
      cursor: 'pointer'
    }
  }, "reset"));
};

var _default = (0, _reactJss.default)(styles)(WalletHeader);

exports.default = _default;
