// oip-react/lib/components/Protobuf/DescriptorSetProto/InputField/InputField.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Creates input field for creating a DescriptorSetProto - Template
const InputField = ({
  classes,
  state,
  onChange,
  onFocus,
  onBlur,
  id,
  type = 'text',
  name = '',
  placeholder = '',
  allowSpaces = true,
  validate
}) => {
  let onKeyDown = () => {};

  if (!allowSpaces) {
    const onChangeCopy = onChange;

    onKeyDown = e => {
      onChange = e.which === 32 ? () => {} : onChangeCopy;
    };
  }

  return /*#__PURE__*/_react.default.createElement("input", {
    required: true,
    type: type,
    id: id,
    name: name,
    value: state[name],
    validate: validate(state[name]),
    onKeyDown: onKeyDown,
    onChange: onChange ? e => onChange(e, id) : null,
    onFocus: onFocus ? e => {
      onFocus(e, id);
    } : null,
    onBlur: onBlur ? e => {
      onBlur(e, id);
    } : null,
    placeholder: placeholder,
    className: (0, _classnames.default)(classes.inputBase, classes.descriptorInputField)
  });
};

var _default = /*#__PURE__*/_react.default.memo(InputField, (oldProps, newProps) => {
  return newProps.shouldUpdate ? newProps.shouldUpdate(oldProps, newProps) : false;
});

exports.default = _default;
