// oip-protobufjs/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ProtoModulesJson", {
  enumerable: true,
  get: function () {
    return _compiled.json;
  }
});
Object.defineProperty(exports, "ProtoModules", {
  enumerable: true,
  get: function () {
    return _compiled.staticModules;
  }
});
Object.defineProperty(exports, "buildOipFiveTemplate", {
  enumerable: true,
  get: function () {
    return _builders.buildOipFiveTemplate;
  }
});
Object.defineProperty(exports, "buildRecordTemplate", {
  enumerable: true,
  get: function () {
    return _builders.buildRecordTemplate;
  }
});
Object.defineProperty(exports, "buildSignedMessage", {
  enumerable: true,
  get: function () {
    return _builders.buildSignedMessage;
  }
});
Object.defineProperty(exports, "signMessage", {
  enumerable: true,
  get: function () {
    return _builders.signMessage;
  }
});
Object.defineProperty(exports, "templateBuilder", {
  enumerable: true,
  get: function () {
    return _builders.templateBuilder;
  }
});
Object.defineProperty(exports, "buildDescriptor", {
  enumerable: true,
  get: function () {
    return _builders.buildDescriptor;
  }
});
Object.defineProperty(exports, "decodeDescriptor", {
  enumerable: true,
  get: function () {
    return _builders.decodeDescriptor;
  }
});
Object.defineProperty(exports, "recordProtoBuilder", {
  enumerable: true,
  get: function () {
    return _builders.recordProtoBuilder;
  }
});
Object.defineProperty(exports, "buildOipDetails", {
  enumerable: true,
  get: function () {
    return _builders.buildOipDetails;
  }
});
Object.defineProperty(exports, "buildTxids", {
  enumerable: true,
  get: function () {
    return _builders.buildTxids;
  }
});
Object.defineProperty(exports, "buildRecord", {
  enumerable: true,
  get: function () {
    return _builders.buildRecord;
  }
});

var _compiled = require("./compiled");

var _builders = require("./builders");
