// oip-react/lib/components/Protobuf/RecordProto/RecordInterface/RecordInterface.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _Publisher = _interopRequireDefault(require("../../../Publisher/Publisher/Publisher"));

var _EnumRow = _interopRequireDefault(require("../EnumRow"));

var _FieldRow = _interopRequireDefault(require("../FieldRow"));

var _RecordProto = _interopRequireDefault(require("../RecordProto"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** * Loops through webFmt object - returns a functional component ***/
const RecordInterface = ({
  classes,
  webFmt,
  dispatch,
  onSuccess,
  onError,
  getMessage,
  mainnetExplorerUrl,
  testnetExplorerUrl,
  withPublisher,
  extendedTemplates,
  oipdHttpApi,
  root,
  setChildState
}) => {
  return /*#__PURE__*/_react.default.createElement("div", {
    className: classes.root
  }, Object.keys(webFmt.fields).map((field, index) => {
    const fieldData = webFmt.fields[field];
    const enumData = webFmt.enums[webFmt.fields[field].enumRefName];

    if (webFmt.fields[field].enumRefName) {
      return /*#__PURE__*/_react.default.createElement(_EnumRow.default, {
        key: `${field}-enum-${index}`,
        enumField: field,
        enumData: enumData,
        classes: classes,
        dispatch: dispatch
      });
    } else {
      return /*#__PURE__*/_react.default.createElement(_FieldRow.default, {
        key: `${field}-${index}`,
        id: `${field}-${index}`,
        field: field,
        fieldData: fieldData,
        classes: classes,
        dispatch: dispatch
      });
    }
  }), Object.keys(extendedTemplates).map((templateIdentifier, i) => {
    const template = extendedTemplates[templateIdentifier];
    return /*#__PURE__*/_react.default.createElement(_RecordProto.default, {
      classes: classes,
      key: templateIdentifier,
      mainnetExplorerUrl: mainnetExplorerUrl,
      testnetExplorerUrl: testnetExplorerUrl,
      oipdHttpApi: oipdHttpApi,
      template: template,
      rootKey: `${root}-${i}`,
      __liftDetails: setChildState
    });
  }), withPublisher && /*#__PURE__*/_react.default.createElement(_Publisher.default, {
    classes: classes,
    onSuccess: onSuccess,
    onError: onError,
    getMessage: getMessage,
    mainnetExplorerUrl: mainnetExplorerUrl,
    testnetExplorerUrl: testnetExplorerUrl
  }));
};

var _default = RecordInterface;
exports.default = _default;
